### ISSUE TITLE
[Bug] bfloat8_b paged_scaled_dot_product_attention_decode PCC regression on Wormhole galaxy (0.999 -> 0.609), caused by #45179 (tilize y/z-dim uninit)

### ISSUE THREAD

**Issue by Aswinmcw** [tenstorrent/tt-metal#46466]
## Summary

`ttnn.transformer.paged_scaled_dot_product_attention_decode` with **`bfloat8_b` K/V** on **Wormhole galaxy (8x4)** regressed from **PCC ~0.9998 to ~0.6088** against an exact paged-attention reference. Bisect pinpoints the first bad commit as:

> **`bb78d08fb43` — "Fix y-dim and z-dim for tilize" (#45179)**, merged 2026-06-08, relates to tt-llk#1161.

This is a real numerics regression on the **row-major-Q decode path** used by production models (Llama3-70B galaxy / Qwen3-32B decode), not a test/golden artifact: the same torch reference hits PCC ~1.0 with `bfloat16` K/V, and reverting only the `#45179` change restores PCC to 0.9998.

## Affected

- Op: `paged_scaled_dot_product_attention_decode` (config_hash `1e561160...`), source `models/demos/llama3_70b_galaxy/tt/llama_attention.py` `forward_decode`.
- Arch: `wormhole_b0`, 6u galaxy (8x4 = 32 chips).
- Trigger: **Q in `ROW_MAJOR_LAYOUT`** (the model's real config) → the decode kernel tilizes Q internally (`tilize_q`, `InitAndUninit`) → hits the changed `_llk_unpack_tilize_uninit_`.
- Only the faithful **`bfloat8_b`** K/V path shows it (bf16 averages over the unpacker corruption).

## git bisect (probe = the standalone repro below, on a 32-chip galaxy)

| commit | description | bf8 PCC |
|---|---|---|
| `0eb131ca2` (HEAD/main) | current | **0.6088 FAIL** |
| `a01cf929674` | parent of culprit | 0.9998 PASS |
| **`bb78d08fb43` (#45179)** | **first bad commit** | **0.6088 FAIL** |
| `97dedae4413` (#45866) | earlier | 0.9998 PASS |

## Root cause

`#45179` changed the Wormhole tilize-uninit to restore the unpacker tile-descriptor **z-dim** to the operand's `num_faces` instead of the previous hardware default `4` (and dropped the y-dim restore):

```c
// tt_metal/tt-llk/tt_llk_wormhole_b0/llk_lib/llk_unpack_tilize.h  (_llk_unpack_tilize_uninit_)
// BEFORE (#45179^, good):
cfg_reg_rmw_tensix<THCON_SEC0_REG0_TileDescriptor_ADDR32 + 1, 16, 0xffff0000>(4);   // z-dim = 4
cfg_reg_rmw_tensix<THCON_SEC0_REG0_TileDescriptor_ADDR32 + 1,  0, 0x0000ffff>(1);   // y-dim = 1

// AFTER (#45179, bad):
cfg_reg_rmw_tensix<THCON_SEC0_REG0_TileDescriptor_ADDR32 + 1, 16, 0xffff0000>(num_faces);  // z-dim = num_faces, y-dim no longer restored
```

For SDPA-decode with row-major Q, the tilized-Q operand's reported `num_faces` is not 4, so the subsequent Q·Kᵀ unpack reads tiles with a wrong z-dim → bf8 attention output collapses.

Isolation (verified on hardware):
- z-dim = `num_faces`, y-dim restored to 1 → still **0.6088** (y-dim is not the issue).
- z-dim = `4`, y-dim = 1 (full revert of the numeric change) → **0.9998** (z-dim is the issue).

## Verified fix

Restoring the z-dim to `4` recovers full PCC (`0.6088 → 0.9998`). The proper fix should land in `_llk_unpack_tilize_uninit_` (revert the z-dim change in `#45179`, or pass/clamp the correct `num_faces` per tt-llk#1161). It's a device-kernel header, so it JIT-recompiles (no host rebuild needed).

## No faithful test-side workaround

The bug only triggers because Q is `ROW_MAJOR_LAYOUT` (the model's traced config). Switching Q to `TILE_LAYOUT` (skips the internal tilize) or K/V to `bfloat16` both avoid it but **change the config hash** and diverge from what the model actually runs — i.e. they hide the regression rather than fix it.

## Reproduction

```bash
TT_METAL_HOME=$PWD PYTHONPATH=$PWD MESH_DEVICE_SHAPE=8x4 \
  ./python_env/bin/python repro_psdpa_decode_8x4.py
# default (faithful bfloat8_b): PCC 0.6088 (FAIL) on main; 0.9998 with the fix
# PSDPA_KV_BF16=1: PCC ~0.9998 (proves the torch reference is correct)
```

<details>
<summary><code>repro_psdpa_decode_8x4.py</code> (place at repo root)</summary>

```python
#!/usr/bin/env python3
# Standalone repro for paged_scaled_dot_product_attention_decode 8x4 config 1e561160
#   source:  models/demos/llama3_70b_galaxy/tt/llama_attention.py (forward_decode)
# Truly mesh-shards (ShardTensor2dMesh) so each chip is an independent, valid paged-decode
# problem with the byte-for-byte traced shapes/placements/dtypes/memory_configs/program_config/
# compute_kernel_config (config_hash unchanged). Per-chip canonical causal paged-attention golden.
#   default (faithful bfloat8_b K/V): PCC 0.6088 on main (regressed by #45179), 0.9998 with fix.
#   PSDPA_KV_BF16=1: PCC ~0.9998 (proves the reference is correct).
import os
import torch
import ttnn

from tests.sweep_framework.master_config_loader_v2 import dict_to_memory_config
from tests.ttnn.utils_for_testing import check_with_pcc

MESH = (8, 4)
SCALE = 0.08838834764831845  # 1/sqrt(128)
SEQ, B, H, D = 1, 8, 8, 128
NKVH = 1
NUM_BLOCKS, BLOCK = 2048, 64
MAX_PAGES = 64

Q_MEM = {
    "buffer_type": "L1", "created_with_nd_shard_spec": False, "memory_layout": "HEIGHT_SHARDED",
    "shard_spec": {"grid": [{"end": {"x": 3, "y": 1}, "start": {"x": 1, "y": 0}}, {"end": {"x": 2, "y": 2}, "start": {"x": 1, "y": 2}}], "orientation": 0, "shape": [8, 128]},
}
OUT_MEM = {
    "buffer_type": "L1", "created_with_nd_shard_spec": False, "memory_layout": "HEIGHT_SHARDED",
    "shard_spec": {"grid": [{"end": {"x": 3, "y": 1}, "start": {"x": 1, "y": 0}}, {"end": {"x": 2, "y": 2}, "start": {"x": 1, "y": 2}}], "orientation": 0, "shape": [32, 128]},
}


def _user_paged_golden(q_user, k_chip, v_chip, page_row, pos, block, scale):
    d = q_user.shape[-1]
    n_active = int(pos) + 1
    n_blocks = (n_active + block - 1) // block
    pages = page_row[:n_blocks].long().clamp_(0, k_chip.shape[0] - 1)
    k_seq = k_chip[pages, 0].reshape(-1, d)[:n_active].float()
    v_seq = v_chip[pages, 0].reshape(-1, d)[:n_active].float()
    w = torch.softmax((q_user.float() @ k_seq.t()) * scale, dim=-1)
    return w @ v_seq


def _open_device():
    ax = os.environ.get("PSDPA_AXIS", "col").lower()
    axis = ttnn.DispatchCoreAxis.ROW if ax == "row" else ttnn.DispatchCoreAxis.COL
    print(f"opening {MESH} mesh, WORKER {ax.upper()} dispatch ...", flush=True)
    return ttnn.open_mesh_device(
        mesh_shape=ttnn.MeshShape(*MESH),
        l1_small_size=79104,
        dispatch_core_config=ttnn.DispatchCoreConfig(ttnn.DispatchCoreType.WORKER, axis),
    )


def main():
    torch.manual_seed(0)
    mesh = _open_device()
    try:
        q_mem = dict_to_memory_config(Q_MEM)
        out_mem = dict_to_memory_config(OUT_MEM)

        q_host = (torch.rand(MESH[0] * SEQ, B, H, D * MESH[1]) * 2 - 1).to(torch.bfloat16)
        k_host = (torch.rand(NUM_BLOCKS, NKVH, BLOCK * MESH[0], D * MESH[1]) * 2 - 1).to(torch.bfloat16)
        v_host = (torch.rand(NUM_BLOCKS, NKVH, BLOCK * MESH[0], D * MESH[1]) * 2 - 1).to(torch.bfloat16)
        GLOBAL_B = B * MESH[1]
        cur_pos_host = torch.tensor([5 + (i * 7) % (BLOCK - 6) for i in range(GLOBAL_B)], dtype=torch.int32)
        page_table_host = (torch.arange(GLOBAL_B * MAX_PAGES, dtype=torch.int32) % NUM_BLOCKS).reshape(GLOBAL_B, MAX_PAGES)

        Q = ttnn.from_torch(
            q_host, dtype=ttnn.bfloat16, layout=ttnn.ROW_MAJOR_LAYOUT, device=mesh,
            mesh_mapper=ttnn.ShardTensor2dMesh(mesh, dims=(0, 3), mesh_shape=MESH), memory_config=q_mem,
        )
        kv_mapper = ttnn.ShardTensor2dMesh(mesh, dims=(2, 3), mesh_shape=MESH)
        kv_dtype = ttnn.bfloat16 if os.environ.get("PSDPA_KV_BF16") else ttnn.bfloat8_b
        K = ttnn.from_torch(k_host, dtype=kv_dtype, layout=ttnn.TILE_LAYOUT, device=mesh,
                            memory_config=ttnn.DRAM_MEMORY_CONFIG, mesh_mapper=kv_mapper)
        V = ttnn.from_torch(v_host, dtype=kv_dtype, layout=ttnn.TILE_LAYOUT, device=mesh,
                            memory_config=ttnn.DRAM_MEMORY_CONFIG, mesh_mapper=kv_mapper)
        PT = ttnn.from_torch(
            page_table_host, dtype=ttnn.int32, layout=ttnn.ROW_MAJOR_LAYOUT, device=mesh,
            memory_config=ttnn.DRAM_MEMORY_CONFIG,
            mesh_mapper=ttnn.ShardTensor2dMesh(mesh, dims=(None, -2), mesh_shape=MESH),
        )
        CP = ttnn.from_torch(
            cur_pos_host, dtype=ttnn.int32, layout=ttnn.ROW_MAJOR_LAYOUT, device=mesh,
            memory_config=ttnn.DRAM_MEMORY_CONFIG,
            mesh_mapper=ttnn.ShardTensor2dMesh(mesh, dims=(None, 0), mesh_shape=MESH),
        )

        _pc_kw = dict(
            compute_with_storage_grid_size=ttnn.CoreCoord(8, 6),
            q_chunk_size=0, k_chunk_size=0, exp_approx_mode=False, max_cores_per_head_batch=16,
        )
        if not os.environ.get("PSDPA_NO_SUBCORE"):
            _pc_kw["sub_core_grids"] = ttnn.CoreRangeSet({
                ttnn.CoreRange(ttnn.CoreCoord(1, 0), ttnn.CoreCoord(3, 9)),
                ttnn.CoreRange(ttnn.CoreCoord(5, 0), ttnn.CoreCoord(6, 8)),
            })
        pc = ttnn.SDPAProgramConfig(**_pc_kw)

        compute_kernel_config = ttnn.WormholeComputeKernelConfig(
            math_fidelity=ttnn.MathFidelity.HiFi2, math_approx_mode=False,
            fp32_dest_acc_en=False, packer_l1_acc=False,
        )

        print("calling paged_scaled_dot_product_attention_decode ...", flush=True)
        out = ttnn.transformer.paged_scaled_dot_product_attention_decode(
            Q, K, V, page_table_tensor=PT, cur_pos_tensor=CP, scale=SCALE, memory_config=out_mem,
            program_config=pc, compute_kernel_config=compute_kernel_config,
        )
        ttnn.synchronize_device(mesh)

        out_dts = ttnn.get_device_tensors(out)
        q_dts = ttnn.get_device_tensors(Q)
        k_dts = ttnn.get_device_tensors(K)
        v_dts = ttnn.get_device_tensors(V)
        pt_dts = ttnn.get_device_tensors(PT)
        cp_dts = ttnn.get_device_tensors(CP)

        o0 = ttnn.to_torch(out_dts[0])
        X, Y, Dd = o0.shape[1], o0.shape[2], o0.shape[3]
        block = ttnn.to_torch(k_dts[0]).shape[2]
        PADDED = 32
        b_eff = max(1, (X * Y) // PADDED)
        stride = max(1, PADDED // Y)

        all_g, all_d = [], []
        for i in range(len(out_dts)):
            ot = ttnn.to_torch(out_dts[i]).float()
            qc = ttnn.to_torch(q_dts[i]).float().reshape(SEQ, B, H, D)
            kc, vc = ttnn.to_torch(k_dts[i]).float(), ttnn.to_torch(v_dts[i]).float()
            cp = ttnn.to_torch(cp_dts[i]).reshape(-1)
            pt = ttnn.to_torch(pt_dts[i]).reshape(B, -1)
            for k in range(b_eff):
                all_g.append(_user_paged_golden(qc[0, k], kc, vc, pt[k], int(cp[k]), block, SCALE))
                all_d.append(ot[0, stride * k, :H, :])
        pcc = check_with_pcc(torch.stack(all_g), torch.stack(all_d), 0.99)
        tag = "PSDPA_KV_BF16" if os.environ.get("PSDPA_KV_BF16") else "faithful bf8"
        print(f"\n==== per-chip paged-attention PCC ({tag}, b_eff={b_eff}/chip over "
              f"{len(out_dts)} chips) = {pcc[1]}  (pass>=0.99: {pcc[0]}) ====", flush=True)
    finally:
        ttnn.close_mesh_device(mesh)


if __name__ == "__main__":
    main()
```

</details>

## Environment

- tt-metal `main` @ `bb78d08fb43`..HEAD (regression present); ttnn `0.65.1rc17`.
- 6u Wormhole galaxy (8x4, 32 chips), TT-KMD 2.4.1.

**Comment by Aswinmcw** [comment]
@ldjurovicTT can u please take a look?
